# Project 4 -- Niharika Rai 

**TA Help:** (for instance) John Smith, Alice Jones, etc., list names of any TAs who helped you

- For example: Help with figuring out how to write a function (describe the tasks that they helped you with)

**Collaboration:** My Friend in CS, My Uncle, Another Student, etc., list names of any other people who helped you

(describe the tasks that they helped you with)
- For example: helped figuring out how to load the dataset.
- Another example: helped debug error with my plot.

**Internet Resources:** Stack Exchange, Stack Overflow, etc.

(describe any information that you learned from internet resources, including the URLs)
- data frames in Pandas versus R from StackOverflow  https://stackoverflow.com/questions/8991709/why-were-pandas-merges-in-python-faster-than-data-table-merges-in-r-in-2012

**ChatGPT, Gemini, Claude, etc:** Any language models or generative AI chatbots that helped you.

(if you used any such tools, please tell us here)
- For example:  I asked ChatGPT how to define a new data frames
- Another example:  Gemini told me how to make a function for sorting my data

**OVERALL MESSAGE:** Any time that you used anything except your brain to solve the questions in these projects, you need to disclose such resources at the start of the project, with details about your usage of the tools.

**YOUR OWN WORK:** Even when you utilize other resources, do NOT just copy and paste.  Write all explanations in your own words, using several sentences in English, which are understandable and which you wrote (and did not just copy and paste).

## Question 1

In [14]:
%%bash 
cd /anvil/projects/tdm/data/restaurant/
pwd

/anvil/projects/tdm/data/restaurant


In [15]:
%%bash 
head /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, '{print $2}'

customer_id
92PEE24
QS68UD8
MB7VY5F
KDJ951Y
BAL0RVT
U263OCD
I9DNSMJ
I9DNSMJ
QYXXJCF


In [16]:
%%bash 
ls /anvil/projects/tdm/data/restaurant | awk '/\.csv$/'

SampleSubmission (1).csv
orders.csv
test_customers.csv
test_full.csv
test_locations.csv
train_customers.csv
train_full.csv
train_locations.csv
vendors.csv


In [21]:
%%bash 
head -n5 /anvil/projects/tdm/data/amazon/tracks.csv | awk -F "|" '{print $2}'

title
"Silent Night"
"Tanssi vaan"
"No One Could Ever"
"Si Vos Querés"


1.2 I followed the example given in the introduction and Dr. Ward's videos to print the second column. 
1.3 I used ChatGPT to get the regex expression ('/\.csv$/') for printing all CSV files.  
1.4 The default field separator for awk is a whitespace (spaces or tabs). I had to use ChatGPT to verify this because I initially thought it was a comma. 

## Question 2

In [24]:
%%bash 
ls /anvil/projects/tdm/data/restaurant/orders.csv 

/anvil/projects/tdm/data/restaurant/orders.csv


In [28]:
%%bash 
head /anvil/projects/tdm/data/restaurant/orders.csv 

akeed_order_id,customer_id,item_count,grand_total,payment_mode,promo_code,vendor_discount_amount,promo_code_discount_percentage,is_favorite,is_rated,vendor_rating,driver_rating,deliverydistance,preparationtime,delivery_time,order_accepted_time,driver_accepted_time,ready_for_pickup_time,picked_up_time,delivered_time,delivery_date,vendor_id,created_at,LOCATION_NUMBER,LOCATION_TYPE,CID X LOC_NUM X VENDOR
163238.0,92PEE24,1.0,7.6,2,,0.0,,,No,,0.0,0.0,,,,,,,,2019-07-31 05:30:00,105,2019-08-01 05:30:16,0,,92PEE24 X 0 X 105
163240.0,QS68UD8,1.0,8.7,1,,0.0,,,No,,0.0,0.0,,,,,,,,2019-07-31 05:30:00,294,2019-08-01 05:31:10,0,Work,QS68UD8 X 0 X 294
163241.0,MB7VY5F,2.0,14.4,1,,0.0,,,No,,0.0,0.0,,,,,,,,2019-07-31 05:30:00,83,2019-08-01 05:31:33,0,,MB7VY5F X 0 X 83
163244.0,KDJ951Y,1.0,7.1,1,,0.0,,,No,,0.0,0.0,,2019-08-01 04:03:00,,,,,,2019-07-31 05:30:00,90,2019-08-01 05:34:54,0,Home,KDJ951Y X 0 X 90
163245.0,BAL0RVT,4.0,27.2,1,,0.0,,,No,,0.0,0.0,,,,,,,,2019-07-31 05:30:00,83,2019-08-01 05:35:51,0,

In [39]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, '{total_item_count += $3} END{print total_item_count}'

308801


In [40]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, '{total_revenue += $3 * $4} END{print total_revenue}'

6.35737e+06


In [42]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, '{total_revenue += $3 * $4} END{printf "%.2f\n", total_revenue}'

6357367.10


In [43]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, '{if ($3 * $4 >= 100) count++} END{print count}'

14183


In [53]:
%%bash 
cut -d, -f2 /anvil/projects/tdm/data/restaurant/orders.csv | sort | uniq -c | sort -n | tail

    107 HX76ORD
    109 BUIAFFT
    113 TVSPG2H
    114 KXXRD9J
    115 2JJAO46
    119 B28LJKM
    128 HJFTTGW
    146 VDEJEMP
    151 TL7Z2DM
    262 XW90EAP


In [54]:
%%bash 
cut -d, -f2 /anvil/projects/tdm/data/restaurant/orders.csv | sort | uniq -c | sort -n | tail -n1

    262 XW90EAP


In [57]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, '{if ($2 == "XW90EAP") {totalspending += $3*$4}} END{print totalspending}'

9180


I used the sample codes and the videos to construct my code. However, I used ChatGPT to learn how to use the printf command and display my answer in 2 decimal places ("%.2f\n") rather than scientific notation.  
2.6 The most frequently ordered customer's ID was XW90EAP, and they made a total of 262 orders.  
2.7 The most frequently ordered customer spent a total of $9180. 

## Question 3

In [58]:
%%bash 
cd /anvil/projects/tdm/data/restaurant/
pwd 

/anvil/projects/tdm/data/restaurant


In [61]:
%%bash 
head -n1 /anvil/projects/tdm/data/restaurant/test_full.csv

customer_id,gender,status_x,verified_x,created_at_x,updated_at_x,location_number,location_type,latitude_x,longitude_x,id,authentication_id,latitude_y,longitude_y,vendor_category_en,vendor_category_id,delivery_charge,serving_distance,is_open,OpeningTime,OpeningTime2,prepration_time,commission,is_akeed_delivering,discount_percentage,status_y,verified_y,rank,language,vendor_rating,sunday_from_time1,sunday_to_time1,sunday_from_time2,sunday_to_time2,monday_from_time1,monday_to_time1,monday_from_time2,monday_to_time2,tuesday_from_time1,tuesday_to_time1,tuesday_from_time2,tuesday_to_time2,wednesday_from_time1,wednesday_to_time1,wednesday_from_time2,wednesday_to_time2,thursday_from_time1,thursday_to_time1,thursday_from_time2,thursday_to_time2,friday_from_time1,friday_to_time1,friday_from_time2,friday_to_time2,saturday_from_time1,saturday_to_time1,saturday_from_time2,saturday_to_time2,primary_tags,open_close_flags,vendor_tag,vendor_tag_name,one_click_vendor,country_id,city_id,created_at_y,updat

In [76]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/test_full.csv | awk -F, 'NR > 1 {count[$2]++} END {for (g in count) print g, count[g]}'

 520000
Male 1062300
Female 89700


In [86]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/test_full.csv | awk -F, 'NR > 1 && $2 != "" {count[$2]++} END {for (g in count) print g, count[g]}'

Male 1062300
Female 89700


In [97]:
%%bash 
head -n1 /anvil/projects/tdm/data/restaurant/orders.csv

akeed_order_id,customer_id,item_count,grand_total,payment_mode,promo_code,vendor_discount_amount,promo_code_discount_percentage,is_favorite,is_rated,vendor_rating,driver_rating,deliverydistance,preparationtime,delivery_time,order_accepted_time,driver_accepted_time,ready_for_pickup_time,picked_up_time,delivered_time,delivery_date,vendor_id,created_at,LOCATION_NUMBER,LOCATION_TYPE,CID X LOC_NUM X VENDOR


In [100]:
%%bash
tail /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, '{print $9}' 

No
No
No
No
No
No
No
No
No
No


In [95]:
%%bash
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, 'NR>1 {if ($9 == "Yes") count++} END{print count}'

1659


In [107]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, 'NR>1 {if ($9 == "Yes") count[$22]++} END {for (v in count) print v, count[v]}' | sort -k2 -nr | head -n1

79 114


I found out that the is_favorite is the 9th column and its rows contain either a "Yes" or "No". The vendor_id is the 22nd column and it contains numbers. 

3.2 The total number of orders for each gender group are: 
Male = 1062300
Female = 89700 
There were also an additional 520000 orders whose rows did not have any gender listed. I have included both outputs. I used ChatGPT to create a more cleaner output (&& $2 != "") that showed only the orders for males and females. 
3.3 A total of 1659 orders were placed from the customer's favorite vendor. I had to count the number of rows where the is_favorite column had a "Yes".  
3.4 Vender number 79 is marked as a favorite by the highest number of customers. The number of customers are 114. I had to identify where the is_favorite column was a "Yes", then count the vendor_id column for these, and print the unique vender_id with their counts. I then had to sort the second column (counts) in descending order and the 1st row of the output was the one that contained the vender_id to the left with its counts on the right. See output above. I used ChatGPT to find out how to sort the second column in descending order since both columns contained numbers (-k2). 

## Question 4

In [108]:
%%bash 
cd /anvil/projects/tdm/data/restaurant/
pwd

/anvil/projects/tdm/data/restaurant


In [109]:
%%bash 
head -n5 /anvil/projects/tdm/data/restaurant/test_full.csv

customer_id,gender,status_x,verified_x,created_at_x,updated_at_x,location_number,location_type,latitude_x,longitude_x,id,authentication_id,latitude_y,longitude_y,vendor_category_en,vendor_category_id,delivery_charge,serving_distance,is_open,OpeningTime,OpeningTime2,prepration_time,commission,is_akeed_delivering,discount_percentage,status_y,verified_y,rank,language,vendor_rating,sunday_from_time1,sunday_to_time1,sunday_from_time2,sunday_to_time2,monday_from_time1,monday_to_time1,monday_from_time2,monday_to_time2,tuesday_from_time1,tuesday_to_time1,tuesday_from_time2,tuesday_to_time2,wednesday_from_time1,wednesday_to_time1,wednesday_from_time2,wednesday_to_time2,thursday_from_time1,thursday_to_time1,thursday_from_time2,thursday_to_time2,friday_from_time1,friday_to_time1,friday_from_time2,friday_to_time2,saturday_from_time1,saturday_to_time1,saturday_from_time2,saturday_to_time2,primary_tags,open_close_flags,vendor_tag,vendor_tag_name,one_click_vendor,country_id,city_id,created_at_y,updat

In [110]:
%%bash 
head /anvil/projects/tdm/data/restaurant/test_full.csv | awk -F, '{print $30}'

vendor_rating
4.4
4.7
4.5
4.5
4.4
4.6
4.3
4.3
4.5


In [1]:
%%bash
awk -F, '
NR>1 && $30 != ""{
    gender = ($2 == "" ? "Blank" : $2)
    vendor = $22
    rating = $30
    if (rating > max[gender]) {
        max[gender] = rating
        top_vendor[gender] = vendor
    }
}
END {
    for (g in max) {
        print g, top_vendor[g], max[g]
    }
}' /anvil/projects/tdm/data/restaurant/test_full.csv

Blank 13.0 4.8
Male 13.0 4.8
Female 13.0 4.8


In [2]:
%%bash 
awk -F, '
NR>1 && $30 != "" {
    gender = ($2 == "" ? "Blank" : $2)
    rating = $30         
    sum[gender] += rating
    count[gender]++
}
END {
    for (g in sum) {
        avg = sum[g] / count[g]
        printf "%s %.2f\n", g, avg
    }
}' /anvil/projects/tdm/data/restaurant/test_full.csv

Blank 4.35
Male 4.35
Female 4.35


In [142]:
%%bash 
head -n1 /anvil/projects/tdm/data/restaurant/orders.csv

akeed_order_id,customer_id,item_count,grand_total,payment_mode,promo_code,vendor_discount_amount,promo_code_discount_percentage,is_favorite,is_rated,vendor_rating,driver_rating,deliverydistance,preparationtime,delivery_time,order_accepted_time,driver_accepted_time,ready_for_pickup_time,picked_up_time,delivered_time,delivery_date,vendor_id,created_at,LOCATION_NUMBER,LOCATION_TYPE,CID X LOC_NUM X VENDOR


In [148]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, 'NR>1 && $7>0 {count[$22]++} END {for (v in count) print v, count[v]}' | sort -k2 -nr | head -n1

148 183


In [155]:
%%bash 
awk -F, -v vendor="148" '
NR>1 && $22 == vendor {
    total++
    if ($7 > 0) discounted++
}
END {
    pct = (discounted / total) * 100
    printf "Vendor: %s\nTotal Orders: %d\nDiscounted Orders: %d\nPercentage Discounted: %.2f%%\n", vendor, total, discounted, pct
}' /anvil/projects/tdm/data/restaurant/orders.csv

Vendor: 148
Total Orders: 1078
Discounted Orders: 183
Percentage Discounted: 16.98%


In [157]:
%%bash 
head /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, '{print$25}'

LOCATION_TYPE

Work

Home
Work
Work
Work
Work
Other


In [160]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/orders.csv | awk -F, 'NR>1 {if ($25 == "Work") count[$22]++} END {for (v in count) print v, count[v]}' | sort -k2 -nr | head -n1

113 989


4.2 The vendor with the vendor_id 13.0 had the highest rating for all gender categories. Their highest rating was 4.8. I first created a blank catgory for rows with no gender listed (I used ChatGPT for this), found the max of the ratings by gender and then found the vendor with the highest rating for each gender. Hence, my output displays the gender category, vendor_id with the max ratings and the max rating given for that vendor.  

Blank: 13.0 
Male: 13.0
Female: 13.0 

4.3 The average vendor rating for all gender categories were the same and they were 4.35. I followed the same process as I did for part 4.2 but this time, I went ahead and calculated the average vendor ratings for each gender. I used ChatGPT to formulate my output to 2 decimal places and assign the blank gender category for the rows with no gender listed in them.   

Blank: 4.35 
Male: 4.35 
Female: 4.35  

4.4 Vendor with vendor_id 148 offered the most discounts, totalling 183. I used the head function to find that the vendor_discount_amount column is in the 7th position of the dataset.  

4.5 Since the vendor_id with 148 got the most discounts, I used this specific number and calculated the total orders, discounted orders and the percentage discounted of the total orders and formulated my output. I used ChatGPT to use the printf command and formulate my output in an organized manner. 

This vendor (148) received a total of 1078 orders and 16.98% of those had a discount applied to them. 

4.6 Vendor with vendor_id 113 received the most orders (989) delivered to work. I first identified the 25th column as the Location_Type and used head to find 'Work' as one of the column values. I used this attribute and applied my filter to calculate the most orders delivered to work by vendor. The vendor_id is the 22nd column.  

## Question 5

In [162]:
%%bash 
head /anvil/projects/tdm/data/restaurant/train_locations.csv

customer_id,location_number,location_type,latitude,longitude
02SFNJH,0,NA,1.6823915384615757,-78.7897371818186
02SFNJH,1,NA,1.6791365450962032,0.7668233053886399
02SFNJH,2,NA,-0.4986483973653315,0.6612414365454651
RU43CXC,0,Home,0.10085325202709132,0.43816546553915037
BDFBPRD,0,NA,2.523125243914389,0.7334637242284848
WMGKW6W,0,NA,12.66341321723571,-1.429919200187407
NDYLK9A,0,NA,-0.20327980769232123,-78.56688511363677
NDYLK9A,1,NA,-0.20437114344229615,0.43559274875681514
PB2B28D,0,Home,2.5756051968780787,0.7082701813272576


In [168]:
%%bash 
cat /anvil/projects/tdm/data/restaurant/train_locations.csv | awk -F, 'NR>1 {count[$1]++} END{for (id in count) print count[id], id}'| sort -nr| head -n5
cat /anvil/projects/tdm/data/restaurant/train_locations.csv | awk -F, 'END {print "Total lines processed:", NR}'

30 4Y0K8NQ
26 NETYQ1C
20 5EP9F62
19 YYKZN0F
19 4UZKXW2
Total lines processed: 59504


In [213]:
%%bash
awk -F, '
    BEGIN {
        print "********************************************";
        print "*            Our Best Customers            *";
        print "********************************************";
    }
    NR > 1 {
        count[$1]++;
    }
    END {
        for (id in count) {
            print count[id], id | "sort -nr | head -n5";
        }
        
        print "********************************************";
        printf"* Total lines processed: %-18s*\n", NR;
        print "********************************************";
    }
' /anvil/projects/tdm/data/restaurant/train_locations.csv

********************************************
*            Our Best Customers            *
********************************************
30 4Y0K8NQ
26 NETYQ1C
20 5EP9F62
19 YYKZN0F
19 4UZKXW2
********************************************
* Total lines processed: 59504             *
********************************************


I first tried to get the customer_id and their counts and find a way to calculate the total lines processed. I then used asterisks and spaces in my code to create a title in the center with asterisks shaped like a border around my output. I used ChatGPT to format my asteriks (%-18s*\n") and ensure that they follow the same structure as shown by the sample output. 

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
